File: compiler/tail_call_test.cpp
#include "vasuki_vm.h"
#include <iostream>
#include <vector>
#include <string>
#include <chrono>

int main() {
    // Create a VM
    vasuki::VasukiVM vm;

    // Create bytecode for a recursive countdown function
    std::vector<uint8_t> bytecode;
    std::vector<vasuki::Value> constants;
    std::vector<std::string> names = {
        "Testing tail call optimization...",
        "countdown",
        "n",
        "Counting down: ",
        "Result: "
    };

    // OpCodes
    const uint8_t HALT = 0;
    const uint8_t PUSH_INT = 2;
    const uint8_t PUSH_STRING = 4;
    const uint8_t GET_LOCAL = 29;
    const uint8_t JUMP = 32;
    const uint8_t JUMP_IF_FALSE = 33;
    const uint8_t CALL = 35;
    const uint8_t TAIL_CALL = 46;
    const uint8_t RETURN = 36;
    const uint8_t FUNCTION = 37;
    const uint8_t GET_GLOBAL = 26;
    const uint8_t PRINT = 42;
    const uint8_t LTE = 20;
    const uint8_t SUB = 11;

    // Push a string
    bytecode.push_back(PUSH_STRING);
    bytecode.push_back(0);
    bytecode.push_back(0);

    // Print the string
    bytecode.push_back(PRINT);

    // Define a countdown function
    bytecode.push_back(FUNCTION);
    bytecode.push_back(1);
    bytecode.push_back(0);
    bytecode.push_back(20);  // Start position
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(1);   // 1 parameter
    bytecode.push_back(2);
    bytecode.push_back(0);   // Parameter name index

    // Jump over function body
    bytecode.push_back(JUMP);
    bytecode.push_back(40);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    // Function body starts here (position 20)
    // Print n
    bytecode.push_back(PUSH_STRING);
    bytecode.push_back(3);
    bytecode.push_back(0);

    bytecode.push_back(PRINT);

    bytecode.push_back(GET_LOCAL);
    bytecode.push_back(2);
    bytecode.push_back(0);

    bytecode.push_back(PRINT);

    // if n <= 0 return 0
    bytecode.push_back(GET_LOCAL);
    bytecode.push_back(2);
    bytecode.push_back(0);

    bytecode.push_back(PUSH_INT);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(LTE);

    bytecode.push_back(JUMP_IF_FALSE);
    bytecode.push_back(10);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    // Return 0
    bytecode.push_back(PUSH_INT);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(RETURN);

    // else return countdown(n-1)
    bytecode.push_back(GET_GLOBAL);
    bytecode.push_back(1);
    bytecode.push_back(0);

    bytecode.push_back(GET_LOCAL);
    bytecode.push_back(2);
    bytecode.push_back(0);

    bytecode.push_back(PUSH_INT);
    bytecode.push_back(1);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(SUB);

    // Call countdown(n-1) with tail call optimization
    bytecode.push_back(TAIL_CALL);
    bytecode.push_back(1);

    bytecode.push_back(RETURN);

    // Main program continues here (position 60)
    // Call countdown(5)
    bytecode.push_back(GET_GLOBAL);
    bytecode.push_back(1);
    bytecode.push_back(0);

    bytecode.push_back(PUSH_INT);
    bytecode.push_back(5);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(CALL);
    bytecode.push_back(1);

    // Print the result
    bytecode.push_back(PUSH_STRING);
    bytecode.push_back(4);
    bytecode.push_back(0);

    bytecode.push_back(PRINT);
    bytecode.push_back(PRINT);

    // Push a value for the main program's implicit return
    bytecode.push_back(PUSH_INT);
    bytecode.push_back(42);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    // Halt
    bytecode.push_back(HALT);

    // Load the bytecode
    vm.loadBytecodeFromMemory(bytecode, constants, names);

    // Execute the bytecode
    std::cout << "Executing bytecode..." << std::endl;

    // First run to define the function
    vm.execute();

    // Reset the bytecode and run again
    bytecode.clear();

    // Call countdown(5)
    bytecode.push_back(PUSH_STRING);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(PRINT);

    bytecode.push_back(GET_GLOBAL);
    bytecode.push_back(1);
    bytecode.push_back(0);

    bytecode.push_back(PUSH_INT);
    bytecode.push_back(5);
    bytecode.push_back(0);
    bytecode.push_back(0);
    bytecode.push_back(0);

    bytecode.push_back(CALL);
    bytecode.push_back(1);

    bytecode.push_back(PUSH_STRING);
    bytecode.push_back(4);
    bytecode.push_back(0);

    bytecode.push_back(PRINT);
    bytecode.push_back(PRINT);

    bytecode.push_back(HALT);

    // Load the new bytecode
    vm.loadBytecodeFromMemory(bytecode, constants, names);

    // Execute the bytecode
    vasuki::Value result = vm.execute();

    // Print the result
    std::cout << "Final result: " << result.toString() << std::endl;

    return 0;
}

File: compiler/tail_call_test.py
#!/usr/bin/env python3
"""
Script to generate bytecode for testing tail call optimization
"""

import struct

# OpCodes
class OpCode:
    HALT = 0
    NOP = 1
    PUSH_INT = 2
    PUSH_FLOAT = 3
    PUSH_STRING = 4
    PUSH_BOOL = 5
    PUSH_NULL = 6
    POP = 7
    POP_N = 8
    DUP = 9
    ADD = 10
    SUB = 11
    MUL = 12
    DIV = 13
    MOD = 14
    NEG = 15
    POW = 16
    EQ = 17
    NEQ = 18
    LT = 19
    LTE = 20
    GT = 21
    GTE = 22
    AND = 23
    OR = 24
    NOT = 25
    GET_GLOBAL = 26
    SET_GLOBAL = 27
    DEFINE_GLOBAL = 28
    GET_LOCAL = 29
    SET_LOCAL = 30
    DEFINE_LOCAL = 31
    JUMP = 32
    JUMP_IF_FALSE = 33
    JUMP_IF_TRUE = 34
    CALL = 35
    RETURN = 36
    FUNCTION = 37
    LIST = 38
    DICT = 39
    GET_PROPERTY = 40
    SET_PROPERTY = 41
    PRINT = 42
    PUSH_CONSTANT = 43
    PUSH_TRUE = 44
    PUSH_FALSE = 45
    TAIL_CALL = 46  # New opcode for tail call optimization

# Generate bytecode for a recursive factorial function with tail call optimization
bytecode = [
    # Push a string
    OpCode.PUSH_STRING, 0, 0,  # String index 0

    # Print the string
    OpCode.PRINT,

    # Define a factorial function with an accumulator for tail call optimization
    OpCode.FUNCTION, 1, 0,  # Function name index 1 (factorial_tail)
    20, 0, 0, 0,  # Start position (20)
    2,            # 2 parameters
    2, 0,         # Parameter 1 name index (2) - n
    3, 0,         # Parameter 2 name index (3) - acc

    # Jump over function body
    OpCode.JUMP, 40, 0, 0, 0,  # Jump to position 60

    # Function body starts here (position 20)
    # if n <= 1 return acc
    OpCode.GET_LOCAL, 2, 0,  # Get parameter n
    OpCode.PUSH_INT, 1, 0, 0, 0,  # Push 1
    OpCode.LTE,              # n <= 1
    OpCode.JUMP_IF_FALSE, 12, 0, 0, 0,  # Jump to else branch if false

    # Return acc
    OpCode.GET_LOCAL, 3, 0,  # Get parameter acc
    OpCode.RETURN,           # Return acc

    # else return factorial_tail(n-1, n*acc)
    OpCode.GET_GLOBAL, 1, 0,  # Get function factorial_tail

    # First argument: n-1
    OpCode.GET_LOCAL, 2, 0,  # Get parameter n
    OpCode.PUSH_INT, 1, 0, 0, 0,  # Push 1
    OpCode.SUB,              # n - 1

    # Second argument: n*acc
    OpCode.GET_LOCAL, 2, 0,  # Get parameter n
    OpCode.GET_LOCAL, 3, 0,  # Get parameter acc
    OpCode.MUL,              # n * acc

    # Call factorial_tail(n-1, n*acc) with tail call optimization
    OpCode.TAIL_CALL, 2,     # Call with 2 arguments (tail call)
    OpCode.RETURN,           # This return is needed for the VM to recognize it as a tail call

    # Main program continues here (position 50)
    # Define a wrapper function that calls factorial_tail with initial accumulator of 1
    OpCode.FUNCTION, 4, 0,  # Function name index 4 (factorial)
    60, 0, 0, 0,  # Start position (60)
    1,            # 1 parameter
    2, 0,         # Parameter 1 name index (2) - n

    # Jump over function body
    OpCode.JUMP, 20, 0, 0, 0,  # Jump to position 80

    # Function body starts here (position 60)
    # return factorial_tail(n, 1)
    OpCode.GET_GLOBAL, 1, 0,  # Get function factorial_tail
    OpCode.GET_LOCAL, 2, 0,   # Get parameter n
    OpCode.PUSH_INT, 1, 0, 0, 0,  # Push 1 (initial accumulator)
    OpCode.CALL, 2,           # Call with 2 arguments
    OpCode.RETURN,            # Return the result

    # Main program continues here (position 75)
    # Print a message
    OpCode.PUSH_STRING, 5, 0,  # String index 5
    OpCode.PRINT,

    # Call factorial(5)
    OpCode.GET_GLOBAL, 4, 0,  # Get function factorial
    OpCode.PUSH_INT, 5, 0, 0, 0,  # Push 5
    OpCode.CALL, 1,           # Call with 1 argument

    # Print the result
    OpCode.PRINT,

    # Print a message
    OpCode.PUSH_STRING, 6, 0,  # String index 6
    OpCode.PRINT,

    # Call factorial(10)
    OpCode.GET_GLOBAL, 4, 0,  # Get function factorial
    OpCode.PUSH_INT, 10, 0, 0, 0,  # Push 10
    OpCode.CALL, 1,           # Call with 1 argument

    # Print the result
    OpCode.PRINT,

    # Print a message
    OpCode.PUSH_STRING, 7, 0,  # String index 7
    OpCode.PRINT,

    # Call factorial(20)
    OpCode.GET_GLOBAL, 4, 0,  # Get function factorial
    OpCode.PUSH_INT, 20, 0, 0, 0,  # Push 20
    OpCode.CALL, 1,           # Call with 1 argument

    # Print the result
    OpCode.PRINT,

    # Push a value for the main program's implicit return
    OpCode.PUSH_INT, 42, 0, 0, 0,  # Push 42

    # Halt
    OpCode.HALT
]

# Constants
constants = []

# Names (strings)
names = [
    "Testing tail call optimization with factorial function...",  # String index 0
    "factorial_tail",                                            # Function name (index 1)
    "n",                                                         # Parameter name (index 2)
    "acc",                                                       # Parameter name (index 3)
    "factorial",                                                 # Function name (index 4)
    "factorial(5) = ",                                           # String index 5
    "factorial(10) = ",                                          # String index 6
    "factorial(20) = "                                           # String index 7
]

# Write bytecode to a file
with open("tail_call_test.bytecode", "wb") as f:
    # Write bytecode
    f.write(struct.pack("<I", len(bytecode)))  # 4-byte unsigned int for size
    for byte in bytecode:
        f.write(struct.pack("B", byte))  # 1-byte unsigned char

    # Write constants
    f.write(struct.pack("<I", len(constants)))  # 4-byte unsigned int for size

    # Write names
    f.write(struct.pack("<I", len(names)))  # 4-byte unsigned int for size
    for name in names:
        encoded = name.encode("utf-8")
        f.write(struct.pack("<I", len(encoded)))  # 4-byte unsigned int for length
        f.write(encoded)

print("Bytecode written to tail_call_test.bytecode")
